ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Iterator.hs

```diff
@@ -503,24 +517,27 @@ iteratorCloseImpl
   :: (HasCallStack, IOLike m)
   => IteratorHandle hash m
   -> m ()
-iteratorCloseImpl IteratorHandle {..} = do
+iteratorCloseImpl IteratorHandle { itState } = do
     atomically (readTVar itState) >>= \case
       -- Already closed
       IteratorStateExhausted -> return ()
-      IteratorStateOpen IteratorState {..} -> do
+      IteratorStateOpen IteratorState { itEpochKey } -> do
         -- First set it to Nothing to indicate it is closed, as the call to
-        -- 'hClose' might fail, which would leave the iterator open in an
+        -- 'release' might fail, which would leave the iterator open in an
         -- invalid state.
         atomically $ writeTVar itState IteratorStateExhausted
-        hClose itEpochHandle
-  where
-    HasFS { hClose } = itHasFS
+        -- NOTE: we must use 'unsafeRelease', as we might be closing the
+        -- iterator from a different thread than the one in which it was
+        -- opened. For example, when closing the ImmutableDB/ChainDB from the
+        -- main thread, we'll close all open iterators, each of them opened in
+        -- a different thread.
+        unsafeRelease itEpochKey
```

Comment:
Comment is wrong (same for other `unsafeRelease`): the reason this was needed is because in the consensus tests, the protocol threads are spawned in a different `ResourceRegistry` than the ChainDB, see #1390. We're not going to fix that in this PR, but add a comment: "this is dangerous, because this allows the iterator to be used by a thread not spawned by this registry, so the iterator can disappear before the thread is terminated".

Reply:
Add a comment on #1390 about this.

Reply:
I'm currently assuming that `iteratorClose` is sometimes called from a different thread than the one that created that registry. E.G. at a glance, it looks like `ChainDB.Impl.implIteratorNext` could be called from a spawned thread, and that calls `ImmDB.iteratorClose`.

So, even if iterators were managed by the ChainDB registry, wouldn't we see `ResourceRegistryUsedFromUnknownThread` without `unsafeRelease`? Or is my assumption wrong?

Reply:
My comment (in the code) was wrong.

A resource can in fact be released from another thread, *as long as that thread has been spawned by the same registry*.

Reply:
Ah ha! I was conflating the two error conditions in just the same way as the comment.

The name `ResourceRegistryUsedFromUnknownThread` is not wrong, but it does admit this misinterpretation. `ResourceRegistryUsedFromUntrackedThread` (or just the full `ResourceRegistryUsedFromThreadNotInRegistry`) would be clearer in this aspect, in hindsight.

Reply:
> The name `ResourceRegistryUsedFromUnknownThread` is not wrong, but it does admit this misinterpretation. `ResourceRegistryUsedFromUntrackedThread` (or just the full `ResourceRegistryUsedFromThreadNotInRegistry`) would be clearer in this aspect, in hindsight.

Agreed. Good idea!

Reply:
I have renamed `ResourceRegistryUsedFromUnknownThread` to `ResourceRegistryUsedFromUntrackedThread` (and updated the comments) in a separate commit.